tasks/batushin_i_striped_matrix_multiplication/mpi/src/ops_mpi.cpp

```diff
@@ -0,0 +1,271 @@
+  const auto &matrix_a = std::get<2>(input);
+
+  const size_t rows_b = std::get<3>(input);
+  const size_t columns_b = std::get<4>(input);
+  const auto &matrix_b = std::get<5>(input);
+
+  if (columns_a != rows_b) {
+    return false;
+  }
+  if (matrix_a.size() != rows_a * columns_a) {
+    return false;
+  }
+  if (matrix_b.size() != rows_b * columns_b) {
+    return false;
+  }
+
+  return true;
+}
+
+bool BatushinIStripedMatrixMultiplicationMPI::PreProcessingImpl() {
+  return ValidationImpl();
+}
+
+namespace {
+
+std::pair<size_t, size_t> CalculateRowDistribution(int rank, int size, size_t n, size_t &rows_per_proc,
+                                                   size_t &extra_rows) {
+  rows_per_proc = n / size;
+  extra_rows = n % size;
+
+  size_t my_rows = rows_per_proc + (std::cmp_less(rank, extra_rows) ? 1 : 0);
+  size_t my_start = (rank * rows_per_proc) + std::min<size_t>(rank, extra_rows);
+
+  return {my_rows, my_start};
+}
+
+void FillLocalPart(size_t my_rows, size_t m, size_t my_start, const std::vector<double> &matrix_a,
+                   std::vector<double> &local_a) {
+  for (size_t i = 0; i < my_rows; ++i) {
+    for (size_t j = 0; j < m; ++j) {
+      local_a[(i * m) + j] = matrix_a[((my_start + i) * m) + j];
+    }
+  }
+}
+
+void FillBufferForProcess(int dest, size_t m, size_t rows_per_proc, size_t extra_rows,
+                          const std::vector<double> &matrix_a, std::vector<double> &buffer) {
+  size_t dest_rows = rows_per_proc + (std::cmp_less(dest, extra_rows) ? 1 : 0);
+  if (dest_rows == 0) {
+    return;
+  }
+
+  size_t dest_start = (dest * rows_per_proc) + std::min<size_t>(dest, extra_rows);
+
+  buffer.resize(dest_rows * m);
+  for (size_t i = 0; i < dest_rows; ++i) {
+    for (size_t j = 0; j < m; ++j) {
+      buffer[(i * m) + j] = matrix_a[((dest_start + i) * m) + j];
+    }
+  }
+}
+
+void SendToProcesses(int size, size_t m, size_t rows_per_proc, size_t extra_rows, const std::vector<double> &matrix_a) {
+  for (int dest = 1; dest < size; ++dest) {
+    std::vector<double> buffer;
+    FillBufferForProcess(dest, m, rows_per_proc, extra_rows, matrix_a, buffer);
+    if (!buffer.empty()) {
+      MPI_Send(buffer.data(), static_cast<int>(buffer.size()), MPI_DOUBLE, dest, 0, MPI_COMM_WORLD);
+    }
+  }
+}
+
+std::vector<double> DistributeMatrixA(int rank, int size, size_t my_rows, size_t my_start, size_t m,
+                                      size_t rows_per_proc, size_t extra_rows, const std::vector<double> &matrix_a) {
+  std::vector<double> local_a(my_rows * m);
+
+  if (rank == 0) {
+    FillLocalPart(my_rows, m, my_start, matrix_a, local_a);
+    SendToProcesses(size, m, rows_per_proc, extra_rows, matrix_a);
+  } else if (my_rows > 0) {
+    MPI_Recv(local_a.data(), static_cast<int>(local_a.size()), MPI_DOUBLE, 0, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
+  }
+
+  return local_a;
+}
+
+std::vector<double> BroadcastMatrixB(int rank, size_t m, size_t p, const std::vector<double> &matrix_b) {
```

Comment:
Matrix B should NOT be broadcasted according to the task description (it should be split vertically and scattered across different ranks)